servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapMerge.java

```diff
@@ -0,0 +1,681 @@
+                                 boolean cancelSubscriberIfNecessary) {
+            final TerminalNotification notification = TerminalNotification.error(throwable);
+            if (trySetTerminal(notification, overrideComplete, terminalNotificationUpdater, this)) {
+                try {
+                    doCancel(cancelSubscriberIfNecessary);
+                } finally {
+                    enqueueAndDrain(notification);
+                }
+                return true;
+            }
+            return false;
+        }
+
+        private void doCancel(boolean cancelSubscription) {
+            Throwable delayedCause = null;
+            // Prevent future onNext operations from adding to subscribers which otherwise may result in
+            // not cancelling mapped Subscriptions.
+            activeMappedSources = Integer.MIN_VALUE;
+            try {
+                if (cancelSubscription) {
+                    assert subscription != null;
+                    subscription.cancel();
+                }
+            } finally {
+                for (FlatMapPublisherSubscriber<T, R> subscriber : subscribers) {
+                    try {
+                        subscriber.cancelFromUpstream();
+                    } catch (Throwable c) {
+                        if (delayedCause == null) {
+                            delayedCause = c;
+                        }
+                    }
+                }
+                subscribers.clear();
+            }
+            if (delayedCause != null) {
+                throwException(delayedCause);
+            }
+        }
+
+        private void enqueueAndDrain(Object item) {
+            if (!pending.offer(item)) {
+                enqueueAndDrainFail(item);
+            }
+            drainPending();
+        }
+
+        private void enqueueAndDrainFail(Object item) {
+            QueueFullException exception = new QueueFullException("pending");
+            if (item instanceof TerminalNotification) {
+                LOGGER.error("Queue should be unbounded, but an offer failed!", exception);
+                throw exception;
+            } else {
+                onError0(exception, true, true);
+            }
+        }
+
+        private boolean acquireEmittingLock() {
+            return acquirePendingLock(emittingLockUpdater, this);
+        }
+
+        private boolean releaseEmittingLock() {
+            return releasePendingLock(emittingLockUpdater, this);
+        }
+
+        private void drainPending() {
+            long emittedCount = 0;
+            for (;;) {
+                if (!acquireEmittingLock()) {
+                    break;
+                }
+                Object t;
+                while ((t = pending.poll()) != null) {
+                    if (sendToTarget(t)) {
+                        ++emittedCount;
+                    }
+                }
+                if (releaseEmittingLock()) {
```

Comment:
wouldn't it be better to call `releaseEmittingLock()` in a finally block ? Just to ensure we can not end up in a limbo state.

Reply:
+1 I'll put in a finally